Add OAuth support for Kafka client in native mode

dejanb opened the pull request.

dejanb committed:
Add OAuth support for Kafka client in native mode

dejanb commented:
This is a first step toward full OAUTH support for the Kafka extension in native mode. These changes get the communication with Keycloak going and obtain a proper token. However, there's an issue with the Kafka client trying to connect to the broker with that token.

I tracked it down to this issue

https://github.com/oracle/graal/issues/2745

Basically,

```Subject subject = Subject.getSubject(AccessController.getContext());```

call return null in the native mode.

I tried to add some classes to the jni configuration, but as with the above reporter that didn't help.

Anyone experienced something like this before? Any pointers on where to look next would be helpful.

Note that the PR in the current state is usable/mergeable, but it doesn't get us to the end.

cescoffier commented on extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaProcessor.java:
```diff
@@ -209,6 +209,30 @@ public void build(CombinedIndexBuildItem indexBuildItem, BuildProducer<Reflectiv
+            reflectiveClass.produce(new ReflectiveClassBuildItem(true, true, true,
+                    "io.strimzi.kafka.oauth.client.JaasClientOauthLoginCallbackHandler",
+                    "org.apache.kafka.common.security.oauthbearer.internals.OAuthBearerRefreshingLogin",
+                    "org.apache.kafka.common.security.oauthbearer.OAuthBearerToken"));
```
Wondering if these 2 last classes shouldn't be always added. WDYT?

dejanb replied:
> About `Subject`, we had a similar issue with Jass and Kafka. We need to check why the lookup fails. IT generally because of a missing class or module.

If you have any pointers on what to try and how to more info please let me know. I can try to provide substitutions for some of these classes and see if that would help.

dejanb replied:
I moved them outside of the strimzi oauth block, so they can be used for general oauth use cases as well.

cescoffier commented on extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaProcessor.java:
```diff
@@ -209,6 +209,30 @@ public void build(CombinedIndexBuildItem indexBuildItem, BuildProducer<Reflectiv
+                    "org.apache.kafka.common.security.oauthbearer.OAuthBearerToken"));
+
+            reflectiveClass.produce(new ReflectiveClassBuildItem(true, true, true,
+                    "org.keycloak.jose.jws.JWSHeader",
```
Does that mean that Strimizi Auth depends on keycloak? How can you be sure that these classes are loadable?

dejanb replied:
Yeah,  Strimzi's `kafka-oauth-client` uses keycloak lib and adds it as a dependency. So it should be available if the library is used.

cescoffier commented:
About `Subject`, we had a similar issue with Jass and Kafka. We need to check why the lookup fails. IT generally because of a missing class or module.

sberyozkin commented:
@dejanb Can `quarkus-oidc` help here ? Why to wire the Keycloak specific code into the Kafka extension ?

dejanb commented:
> About `Subject`, we had a similar issue with Jass and Kafka. We need to check why the lookup fails. IT generally because of a missing class or module.

If you have any pointers on what to try and how to more info please let me know. I can try to provide substitutions for some of these classes and see if that would help.

dejanb commented:
> @dejanb Can `quarkus-oidc` help here ? Why to wire the Keycloak specific code into the Kafka extension ?

@sberyozkin these are used by strimzi oauth client, so this PR is intended to provide a better support for that case. I'd need to give oidc a look, but I assume that the client would need first changed to use it.

sberyozkin commented:
@dejanb I see, so in your PR is only about getting a token by directly requesting it from Keycloak. So `quarkus-oidc` won't help at the moment, I've started working on `OidcClient` (`quarkus-oidc-client`) which will work not only with Keycloak, but it is not ready yet.
Thanks

dejanb commented:
@sberyozkin Great. We can take a look of moving strimzi oauth to it later on.

cescoffier commented:
@dejanb Sorry, missed the notifications.

I would go with simple substitutions adding traces to check where it fails.

dejanb commented:
@cescoffier no worries, I was on other stuff as well. I pretty much pin-pointed the problematic places already, I'll try to play with substitutions to come up with a fix. Hopefully, I'll have something to show this week. Thanks for all your help so far.

cescoffier commented:
Can you link the place you think it fails?

dejanb commented:
> Can you link the place you think it fails?
@cescoffier 

So what I think is going on is that `SaslClientAuthenticator` do `Subject.doAs()` call with the proper subject it wants to use

https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslClientAuthenticator.java#L524

This subject later on should be available with the call to `Subject.getSubject(AccessController.getContext())`, but this call returns `null` in the callback handler.

https://github.com/apache/kafka/blob/aa287acb2eed07cf6d75c10e71f051f538a57872/clients/src/main/java/org/apache/kafka/common/security/oauthbearer/internals/OAuthBearerSaslClientCallbackHandler.java#L99

That's where it fails.

Let me know if this helpful or do you need any more info from my side.

cescoffier commented:
IT seems close to the SASL issue we had with kafka a few months ago.

dejanb commented:
@cescoffier unfortunately SASL fix wasn't applicable here. I ended up substituting `Subject` class which would help with all these `doAs()` -> `getSubject()` usage patterns until graalvm issue is fixed. It might not be ideal but all other solution would need much more substitutions to avoid `Subject` usage at all. This now works nicely. Let me know what you think.

cescoffier commented:
Thanks @dejanb. 
@galderz WDYT about the substitutions?

cescoffier commented on extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/graal/SubstitutionUtils.java:
```diff
@@ -0,0 +1,12 @@
+
+import javax.security.auth.Subject;
+
+public class SubstitutionUtils {
```
Not sure about the name.
Maybe `SubjectHolder` ?

cescoffier commented on extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/graal/SubstitutionUtils.java:
```diff
@@ -0,0 +1,12 @@
+
+public class SubstitutionUtils {
+
+    public static HashMap<AccessControlContext, Subject> subjects = new HashMap<>();
```
Should it be `final`?

cescoffier replied:
Should this field be moved to the `Subject` class?

galderz replied:
@cescoffier The field has to be kept outside otherwise IIRC you get an error. The only fields allowed in substituted classes are aliases of existing fields.

I'm not an expert on this java security APIs, but a couple of things to bear in mind would be:

1. Should be it be concurrent?
2. Is there a chance entries could be stored and not removed? Maybe you need some kind of expiration of entries to avoid a leak?

cescoffier replied:
ah, thanks for reminding me of this @galderz! I always forgot. 

Handling expiration might be good as it would avoid the potential growth I mentioned in my review.

cescoffier reviewed (changes requested):
Nice trick. 
I've done a few suggestions around SubstitutionsUtils.

There is a potential memory leak if there is lots of call to `doAs` breaking before the call to `getSubject`- the map size could grow unexpectedly. 
I guess it's fine here, as we won't have such a scenario.

dejanb commented:
> Nice trick.
> I've done a few suggestions around SubstitutionsUtils.
> 
> There is a potential memory leak if there is lots of call to `doAs` breaking before the call to `getSubject`- the map size could grow unexpectedly.
> I guess it's fine here, as we won't have such a scenario.

@cescoffier thanks, great suggestions. I'll incorporate them in next.

dejanb commented:
@cescoffier @galderz Thanks for the feedback. I pushed some more changes. As Galder mentioned, the static field cannot be added to the target class. But I moved the `SubjectHolder` class to the target class and that works and looks a bit better.

In terms of actual map, I made it synchronized and converted to `WeakHashMap()` which removes entries when the key is no longer in use. I think this should help with potential leaks. 

I can use full blown cache (with expiring entries) here but I'll need to add a dependency on Infinispan/Guava for that. Let me know if you want me to explore that route.

galderz commented:
@dejanb Rather than `WeakHashMap`, I'd probably use `Caffeine` for this which is a library already in use within Quarkus in different places (e.g. to power caching for quarkus end user apps). I'm not expert on WeakHashMap but you can find some details [here](https://stackoverflow.com/questions/1802809/javas-weakhashmap-and-caching-why-is-it-referencing-the-keys-not-the-values#:~:text=4%20Answers&text=WeakHashMap%20isn't%20useful%20as,use%20it%20for%2C%20incorrectly) on the gotchas of using them as a cache. Also, I don't think `WeakHashMap` is designed for concurrent access.

dejanb commented:
@galderz Makes sense. I pushed an implementation using Caffeine cache. I used some defaults for sizes and expiry times we use on other projects and it seems to me they should work well here as well. Let me know if you have any other suggestions.

machi1990 commented on extensions/kafka-client/runtime/pom.xml:
```diff
@@ -34,6 +34,11 @@
 
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-caffeine-deployment</artifactId>
```
```suggestion
            <artifactId>quarkus-caffeine</artifactId>
```

We should depend on the runtime module and not the `deployment` one.

dejanb replied:
Thanks for the suggestion. I fixed it, however I needed to add a dependency on this deployment module in `kafka-client-deployment` to make it work properly.

machi1990 replied:
Thanks @dejanb. The CI failure seemed unrelated to change, I've re-launched the build.

cescoffier commented:
Re-running the CI before merging.

cescoffier commented:
@dejanb Sorry for the delay. It should be fine to merge it, but I re-run the CI to be sure.

machi1990 commented:
This errored again with
```
Error: Unable to process command '::set-env name=JAVA_HOME_11_x64::/opt/hostedtoolcache/AdoptOpenJDK/1.0.0-ga-11-jdk-hotspot-linux-x64-normal-latest/x64' successfully.
Error: The `set-env` command is disabled. Please upgrade to using Environment Files or opt into unsecure command execution by setting the `ACTIONS_ALLOW_UNSECURE_COMMANDS` environment variable to `true`. For more information see: 
```

I've this these error before (a month or so ago) and they were fixed by @loicmathieu or @gsmet. I think a rebase will do and it will re-trigger a clean build. Do you want to take a look? @dejanb

dejanb commented:
@cescoffier @machi1990 Done. Hopefully it'll do it.

cescoffier commented:
@dejanb It's IN! What a great way to start the year!

dejanb commented:
@cescoffier awesome ... To many more in this year :)
